core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -390,9 +394,9 @@ public void addTaskDynamicFilters(TaskId taskId, Map<DynamicFilterId, Domain> ne
             return;
         }
         checkState(
-                taskAttemptId == context.getAttemptId(),
-                "Task %s retry attempt %s has not been registered with dynamic filter service",
-                taskId,
+                taskRetriesEnabled(context) || taskAttemptId == context.getAttemptId(),
```

Comment:
Somehow the interactions in this class seem to be unnecessarily complex to what it does. In nutshell the responsiblity of the `DynamicFilterContext` is to collect dynamic filters information from all expected tasks, generate a final domain and publish it. I would consider revisiting the design of this class to make it more readable.

Reply:
I agree that the code in this class is tricky. However, I'm preferring to not redo the code of this class until there is a more pressing need for it.